### PR TITLE
Fix MinGW build

### PR DESCRIPTION
**Pull request by twasilczyk**

### CODE REVIEW

File: mDNSCore/mDNSEmbeddedAPI.h [hunk 1]
```diff
@@ -53,6 +53,11 @@
 #ifndef __mDNSEmbeddedAPI_h
 #define __mDNSEmbeddedAPI_h
 
+// MinGW defines "#define interface struct" for ObjC compatibility.
```
Comment (Abhayakara) [inline mDNSCore/mDNSEmbeddedAPI.h:56]: Is there a preprocessor predefine that's specific to MinGW? We also build in Visual Studio, which I think defines _WIN32 as well.

Comment (Abhayakara) [reply mDNSCore/mDNSEmbeddedAPI.h:56]: Oh, I see in a later change that you're checking for __MINGW32__. Why not do that here as well?

File: mDNSWindows/SystemService/Service.c [hunk 2]
```diff
@@ -29,7 +29,7 @@
 #include	"uds_daemon.h"
 #include	"GenLinkedList.h"
 #include	"Service.h"
-#include	"EventLog.h"
+#include	"mDNSWindows/SystemService/EventLog.h"
```
Comment (Abhayakara) [inline mDNSWindows/SystemService/Service.c:32]: This should be #ifdef __MINGW__ ... #else ... #endif, so that it builds on Visual Studio as well. Better still, fix the MinGW Makefile to add a "-I mDNSWindows/SystemService" so we don't need the ifdef.

File: mDNSWindows/mDNSWin32.c [hunk 3]
```diff
@@ -1054,7 +1054,7 @@ mDNSPlatformTCPConnect
 
 	// Setup connection data object
 
-	sock->userCallback	= inCallback;
+	sock->userCallback	= (void*) inCallback;
```
Comment (Abhayakara) [inline mDNSWindows/mDNSWin32.c:1057]: Why isn't TCPUserCallback of type TCPConnectionCallback?

File: mDNSWindows/mDNSWin32.c [hunk 4]
```diff
@@ -1947,9 +1947,7 @@ SetDomainFromDHCP( void )
 
 	for ( pAdapter = pAdapterInfo; pAdapter; pAdapter = pAdapter->Next )
 	{
-		if ( pAdapter->IpAddressList.IpAddress.String &&
```
Comment (Abhayakara) [inline mDNSWindows/mDNSWin32.c:1950]: What's the warning here? Why would we not want to check the pointer for NULL?

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1950]: It's not a pointer, but a [C array](https://docs.microsoft.com/en-us/windows/win32/api/iptypes/ns-iptypes-ip_address_string).

`warning: address of array 'pAdapter->IpAddressList.IpAddress.String' will always evaluate to 'true' [-Wpointer-bool-conversion]`

File: mDNSWindows/mDNSWin32.c (hunk 3 above)
Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1057]: It has different arguments.

`warning: incompatible function pointer types assigning to 'TCPUserCallback' (aka 'void (*)()') from 'TCPConnectionCallback' (aka 'void (*)(struct TCPSocket_struct *, void *, unsigned char, int)') [-Wincompatible-function-pointer-types]`

File: mDNSWindows/SystemService/Service.c (hunk 2 above)
Comment (twasilczyk) [reply mDNSWindows/SystemService/Service.c:32]: This might be Android buildsystem related issue. Let's leave it out for now.

File: mDNSCore/mDNSEmbeddedAPI.h (hunk 1 above)
Comment (twasilczyk) [reply mDNSCore/mDNSEmbeddedAPI.h:56]: Yes, it's [specific](https://github.com/mirror/mingw-w64/blob/master/mingw-w64-headers/include/rpc.h) to MinGW. I will change the condition.

File: mDNSWindows/mDNSWin32.c (hunk 3 above)
Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c:1057]: Right, but isn't that bad? Shouldn't it have the same arguments?

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1057]: Isn't it a C-style way of passing user callbacks? I'm pretty sure this was exactly the way it was done in other C open source project I worked on.

So [TCPUserCallback](https://github.com/Abhayakara/mdnsresponder/blob/50902620f335639ffc86d0bc41c101e684390f70/mDNSWindows/mDNSWin32.h#L35) is just a function pointer equivalent to `void*` and [TCPConnectionCallback](https://github.com/Abhayakara/mdnsresponder/blob/50902620f335639ffc86d0bc41c101e684390f70/mDNSCore/mDNSEmbeddedAPI.h#L3050) is a concrete callback implementation.

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c]: Having said that, I should have casted it to that pointer type, not to `void*`. Pushed an updated change that matches what's done in the other place (line 1223, function `TCPSocketNotification`).

File: mDNSWindows/Poll.c [hunk 5]
```diff
@@ -113,7 +113,7 @@ ShiftDown( void * arr, size_t arraySize, size_t itemSize, int index )
 
 #define	DEBUG_NAME	"[mDNSWin32] "
 #define gMDNSRecord mDNSStorage
-mDNSlocal Poll gPoll = { mDNSfalse, NULL };
+mDNSlocal Poll gPoll = {};
```
Comment (Abhayakara) [inline mDNSWindows/Poll.c:116]: Why is this change necessary?

File: mDNSWindows/mDNSWin32.c [hunk 6]
```diff
@@ -1054,7 +1054,7 @@ mDNSPlatformTCPConnect
 
 	// Setup connection data object
 
-	sock->userCallback	= inCallback;
+	sock->userCallback	= (TCPUserCallback) inCallback;
```
Comment (Abhayakara) [inline mDNSWindows/mDNSWin32.c:1057]: Your response to my question about this was to say that it should be cast to the right type pointer, but that's not right. The callback should be of the right type. There shouldn't be a cast here at all. This is not safe coding.

File: mDNSWindows/Poll.c (hunk 5 above)
Comment (twasilczyk) [reply mDNSWindows/Poll.c:116]: Because it initializes only the first two fields of `struct Poll_struct` / `Poll`. The changed line initializes the whole struct. I don't remember the exact warning text, but I can check it for you if you're curious.

File: mDNSWindows/mDNSWin32.c (hunk 6 above)
Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1057]: Not precisely a "right" pointer type, my point was primarily saying this is a wacky way to simulate lambdas in C. And it was just my guess of intentions of whoever wrote this code.

Having said that, I'm happy to change it to the correct pointer type. Indeed after some more in-depth code searching it looks like this specific callback isn't used the way I assumed it is (i.e. like simulated lambdas) and it should work to just replace TCPUserCallback with TCPConnectionCallback. I'll jump on it as soon as you give me green light on all other changes here :). I would really like to see at least some of them merged.

File: mDNSWindows/Poll.c (hunk 5 above)
Comment (Abhayakara) [reply mDNSWindows/Poll.c:116]: Ah, okay, that explains it. I don't see any issue with this—just wanted a reason to mention. :)

File: mDNSWindows/mDNSWin32.c (hunk 6 above)
Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c:1057]: Sounds good. I too would like to get this stuff merged—sorry it's taken so long!

Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c:1057]: Am I waiting on you for this change, or did I drop the ball somewhere? :)

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1057]: This time it's on me - I misunderstood one of your previous comments, sorry!

I tried making the change we discussed, but it turned out the callback is indeed used as a quirky C-style lambda in two places with two different argument sets: `mDNSWindows/SystemService/Service.c` and `mDNSWindows/mDNSWin32.c`. We could use an union here to avoid casting (PTAL how I did it on top of this change: https://pastebin.com/Dgyunh9F). Please let me know if that's what you're looking for so I would apply the linked diff on top of this change.

I know it would be best to use discriminated unions (i.e. `std::variant`) or just lambdas/bind expressions (i.e. `std::function`), but this project is C-only, I'm afraid :).

File: mDNSWindows/mDNSWin32.c (hunk 3 above)
Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c]: Why not just delete this:

mDNSWin32.h:typedef void ( *TCPUserCallback )();

and change this:

mDNSWin32.h:	TCPUserCallback				userCallback;
 
to this:

mDNSWin32.h:	TCPConnectionCallback				userCallback;

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c]: This is exactly what I tried first, but then I got build error:

```
external/mdnsresponder/mDNSWindows/SystemService/Service.c:2023:68: error: too few arguments to function call, expected 4, have 2
                tcpSock->userCallback( ( int ) tcpSock->fd, tcpSock->userContext );
                ~~~~~~~~~~~~~~~~~~~~~                                            ^
external/mdnsresponder/mDNSWindows/SystemService/Service.c:2050:23: warning: incompatible function pointer types assigning to 'TCPConnectionCallback' (aka 'void (*)(struct TCPSocket_struct *, void *, unsigned char, int)') from 'udsEventCallback' (aka 'void (*)(int, void *)') [-Wincompatible-function-pointer-types]
                sock->userCallback              = callback;
                                                ^ ~~~~~~~~
```

Because of what I described in the other comment thread (why there is two btw?).

File: mDNSWindows/mDNSWin32.c (hunk 6 above)
Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c:1057]: Sorry for my confused comment on an earlier discussion. I ran into this overloading when I updated mDNSPosix to support TCP a few years ago. There is actually (as far as I can see) no reason to do this overloading. These are two different callbacks with two different function signatures. They aren't interchangeable. So the fix is easy: have two different callbacks in the TCPSocket structure.

In the Posix code (and the MacOSX code) uds_daemon uses a unix domain socket, not a TCP socket, so we don't see this conflict. You could do this in the Windows code as well—the fact that both of these structures are managing TCP connections doesn't mean there can only be one structure.

I would offer to just fix this myself, but I don't have a Windows machine to test on. Would you mind trying one of these solutions?

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c:1057]: Makes sense. I started with using separate callbacks in the first place, but then used union to avoid any potential ABI incompatibilities. So the final change will be very similar to https://pastebin.com/Dgyunh9F except the two callbacks will not be grouped into an union. Let me push the updated change shortly.

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c]: Pushed, PTAL.

Comment (Abhayakara) [reply mDNSWindows/mDNSWin32.c]: Great, thanks! I've integrated the other pull requests in our repo, so this is the last one, and then I can sync to the github repo.

Comment (twasilczyk) [reply mDNSWindows/mDNSWin32.c]: Thanks, when you have a chance, PTAL at the spreadsheet I linked at https://github.com/Abhayakara/mdnsresponder/pull/14 for the potential next batch of patches. There's a few more or less obvious bugs and a couple of feature requests.

**Comment by Abhayakara** [comment]
BTW, just so you know, I am still making forward progress on this. The problem is that we are in the middle of a release, and so there have been a few changes that require additional work on my end unrelated to your changes in order to actually do the release that has your changes in it. :]

**Comment by Abhayakara** [comment]
This pull request has been merged to the internal repo and was included in the mDNSResponder-1587 commit.
